pkg/crds/crds.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2023 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package crds
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/rook/kubectl-rook-ceph/pkg/k8sutil"
+	"github.com/rook/kubectl-rook-ceph/pkg/logging"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"strings"
+	"time"
+)
+
+var cephResources = []string{
+	"cephclusters",
+	"cephblockpoolradosnamespaces",
+	"cephblockpools",
+	"cephbucketnotifications",
+	"cephbuckettopics",
+	"cephclients",
+	"cephcosidrivers",
+	"cephfilesystemmirrors",
+	"cephfilesystems",
+	"cephfilesystemsubvolumegroup",
+	"cephnfses",
+	"cephobjectrealms",
+	"cephobjectstores",
+	"cephobjectstoreusers",
+	"cephobjectzonegroups",
+	"cephobjectzones",
+	"cephrbdmirrors",
+}
+
+const (
+	cephRookIoGroup          = "ceph.rook.io"
+	cephRookResourcesVersion = "v1"
+)
+
+const (
+	CephResourceKindCephCluster = "CephCluster"
+	toolBoxDeployment           = "rook-ceph-tools"
+	rookOperatorDeployment      = "rook-ceph-operator"
+	timeOutCleanupResources     = 10 * time.Minute
+	timeOutCheckCleanup         = 15 * time.Second
+)
+
+var (
+	clusterResourcePatchFinalizer = map[string]interface{}{
+		"spec": map[string]interface{}{
+			"cleanupPolicy": map[string]string{
+				"confirmation": "yes-really-destroy-data",
+			},
+		},
+	}
+
+	defaultResourceRemoveFinalizers = map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"finalizers": nil,
+		},
+	}
+)
+
+func DeleteCustomResources(ctx context.Context, clientsets k8sutil.ClientsetsInterface, k8sClientSet kubernetes.Interface, clusterNamespace string) {
+	err := deleteCustomResources(ctx, clientsets, clusterNamespace)
+	if err != nil {
+		logging.Fatal(err)
+	}
+	k8sutil.DeleteDeployment(ctx, k8sClientSet, clusterNamespace, toolBoxDeployment)
+	ensureClusterIsEmpty(ctx, k8sClientSet, clusterNamespace)
```

Comment:
After the deployments are cleaned up, is the tool done? I wonder if we need to also wait for the cleanup job(s) to complete.

Reply:
during my tests seems the resources goes off after a couple of seconds when the process finished, probably would be worth add another 15 seconds of timeout when `ensureClusterIsEmpty()` function  is started.

something like this:
```go
func ensureClusterIsEmpty(ctx context.Context, k8sClientSet kubernetes.Interface, clusterNamespace string) {
	timer := time.NewTimer(timeOutCleanupResources)
       // ensure the checker start after the first 15 seconds
        time.Sleep(timeOutCheckCleanup)
	for {
		deployments, err := k8sClientSet.AppsV1().Deployments(clusterNamespace).List(ctx, v1.ListOptions{LabelSelector: "rook_cluster" + clusterNamespace})
		if err != nil {
			logging.Fatal(err)
		}
 .... //rest of the code
```

What do you think?

Reply:
Why 15 seconds? It's good to avoid sleeps if we can, unless they are a short sleep just to retry to see if something is cleaned up.

Reply:
I just tested the latest, it didn't wait for all the daemon pods to go away and for the cleanup job to start. Even if we just wait for one cleanup job to start and complete, it will help the reliability, thanks!